### DISCUSSION TITLE
Fastlane + Gitlab + Keychains issue

### DISCUSSION THREAD

**Discussion by KacperCzapp** [fastlane/fastlane#18943] [Q&A]
I am running my CI/CD on a Gitlab instance and luckily I have access to the build machine so I can better see the issue.

In my Fastfile I have:
```ruby
  desc "Creates temporary keychain with name specified in `KEYCHAIN_NAME` environment variable"
  lane :create_temp_keychain do

    delete_temp_keychain

    create_keychain(
      unlock: true,
      default_keychain: false,
      timeout: 7200 #2 hours
    )

    UI.message("Keychain path sharedValue: #{lane_context[SharedValues::KEYCHAIN_PATH]}")
  end

  desc "Deletes the temporary keychain with name specified in `KEYCHAIN_NAME` environment variable"
  lane :delete_temp_keychain do
    name = ENV['KEYCHAIN_NAME']
    begin
      delete_keychain
      UI.success("Deleted keychain #{name}")
    rescue => ex
      UI.important("Deleting keychain #{name} failed")
    end
  end
  
  desc "Runs `match`, which downloads the certificates, adds them to the temp keychain
  and sets the code signing to manual with profiles taken from `match`."
  private_lane :update_signing do

    UI.header "Update signing"

    match(
      shallow_clone: true,
      clone_branch_directly: true,
      readonly: true
    )

    appId = ENV['APP_ID']
    matchType = ENV['MATCH_TYPE']
    team_id = ENV["sigh_#{appId}_#{matchType}_team-id"]
    profile_uuid = ENV["sigh_#{appId}_#{matchType}"]

    update_code_signing_settings(
      use_automatic_signing: false,
      team_id: ENV["sigh_#{appId}_#{matchType}_team-id"],
      profile_uuid: ENV["sigh_#{appId}_#{matchType}"],
      profile_name: ENV["sigh_#{appId}_#{matchType}_profile-name"]
    )
  end
```

TL:DR; Each Gitlab pipeline creates a keychain specifically for this pipeline and **should** remove it after the job is done. If it's cancelled, it doesn't unfortunately, meaning there might be leftover keychains for some time (they are cleaned once a day at midnight).

Gitlab has this disadvantage, that it may run several pipelines at the same time. That is, several keychains with the same certificates can be present on the same machine. 

1. Job A creates keychain A
2. Gitlab switches to Job B
3. Job B creates keychain B and locks the keychain A
4. Job B finishes, Gitlab switches back to Job A
5. Job A wants to use the keychain and fails

Another scenario:
1. Job A creates keychain A
2. `match` adds certificates to keychain A
2. Job A is cancelled an keychain A stays on the machine
3. Job B creates keychain B and locks the keychain A (or timeout has passed)
4. `match` adds the same certificates to keychain A
5. CodeSign sees the required certificates on keychain A instead of using B (probably because it's higher on the list of keychains)
6. Job B wants to use the keychain A and asks for password (visible on computer's screen, not via console), which times out and job fails.

I am not able to prevent creation of multiple keychains at once and I also cannot clear the leftovers before a job starts because it might not have been used by another job yet.

Each job has its own keychain so that only currently needed certificates and provisioning profiles are stored on the machine at any given moment. 

Do you have any idea why the locking of other keychains happens or how else to solve this problem?

**Comment by lucgrabowski** [comment]
Hi, I haven't used `create_keychain`, but I have some questions that may help.
Could you pass `keychain_name` to `match` call (list of parameters in https://docs.fastlane.tools/actions/match/#match)? 
Does `ENV["KEYCHAIN_NAME"]` have the same value when `create_keychain` and `update_signing` are run?
There is also `unlock_keychain` action that may be helpful http://docs.fastlane.tools/actions/unlock_keychain/#unlock_keychain

**Reply by KacperCzapp** [reply]
Yes, the keychain names (`MATCH_KEYCHAIN_NAME` and `KEYCHAIN_NAME`) are set up in environment variables of `.gitlab-ci.yml` and both consist of the project name and the commit hash: `$CI_PROJECT_NAME-$CI_COMMIT_SHA`.

**Reply by lucgrabowski** [reply]
I understand that proper values are in `match`'s summary table in logs. Are they?

**Reply by KacperCzapp** [reply]
Yes, though I suppose `match` itself is not the issue. It's `create_keychain` (or maybe `CodeSign`?) that locks all other custom keychains when creating a new one. And then CodeSign tries to access the locked keychain because it sees that the required certificate is in that keychain, instead of the current one created for this specific job.

**Reply by lucgrabowski** [reply]
Can you see if `unlock_keychain` helps?

**Reply by jChris85** [reply]
I am having the same problem.

@KacperCzapp thanks for explaining that the keychains are leftovers from failed Gitlab job. I was wondering where all the keychain files are coming from.

Could you solve the problem?
